staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go

```diff
@@ -0,0 +1,120 @@
+
+// New creates a new table convertor for the provided OpenAPI schema. If the printer definition cannot be parsed,
+// error will be returned along with a default table convertor.
+func New(extensions spec.Extensions) (rest.TableConvertor, error) {
+	headers := []metav1beta1.TableColumnDefinition{
+		{Name: "Name", Type: "string", Format: "name", Description: swaggerMetadataDescriptions["name"]},
+		{Name: "Created At", Type: "date", Description: swaggerMetadataDescriptions["creationTimestamp"]},
+	}
+	c := &convertor{
+		headers: headers,
+	}
+	format, ok := extensions.GetString(printColumnsKey)
+	if !ok {
+		return c, nil
+	}
+	// "x-kubernetes-print-columns": "custom-columns=NAME:.metadata.name,RSRC:.metadata.resourceVersion"
+	parts := strings.SplitN(format, "=", 2)
+	if len(parts) != 2 || parts[0] != "custom-columns" {
+		return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns', only support 'custom-columns=NAME=JSONPATH[,NAME=JSONPATH]'")
+	}
+	columnSpecs := strings.Split(parts[1], ",")
+	var columns []*jsonpath.JSONPath
+	for _, spec := range columnSpecs {
+		parts := strings.SplitN(spec, ":", 2)
+		if len(parts) != 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
+			return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns', must specify NAME=JSONPATH: %s", spec)
+		}
+		path := jsonpath.New(parts[0])
+		if err := path.Parse(parts[1]); err != nil {
+			return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns': %v", spec)
+		}
+		path.AllowMissingKeys(true)
+		columns = append(columns, path)
+		headers = append(headers, metav1beta1.TableColumnDefinition{
+			Name:        parts[0],
+			Type:        "string",
+			Description: fmt.Sprintf("Custom resource definition column from OpenAPI (in JSONPath format): %s", parts[1]),
+		})
+	}
+	c.columns = columns
+	c.headers = headers
+	return c, nil
+}
+
+type convertor struct {
+	headers []metav1beta1.TableColumnDefinition
+	columns []*jsonpath.JSONPath
+}
+
+func (c *convertor) ConvertToTable(ctx genericapirequest.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1beta1.Table, error) {
+	table := &metav1beta1.Table{
+		ColumnDefinitions: c.headers,
+	}
+	if m, err := meta.ListAccessor(obj); err == nil {
+		table.ResourceVersion = m.GetResourceVersion()
+		table.SelfLink = m.GetSelfLink()
+		table.Continue = m.GetContinue()
+	} else {
+		if m, err := meta.CommonAccessor(obj); err == nil {
+			table.ResourceVersion = m.GetResourceVersion()
+			table.SelfLink = m.GetSelfLink()
+		}
+	}
+
+	var err error
+	buf := &bytes.Buffer{}
+	table.Rows, err = metatable.MetaToTableRow(obj, func(obj runtime.Object, m metav1.Object, name, age string) ([]interface{}, error) {
+		cells := make([]interface{}, 2, 2+len(c.columns))
+		cells[0] = name
+		cells[1] = age
+		for _, column := range c.columns {
+			if err := column.Execute(buf, obj); err != nil {
+				cells = append(cells, nil)
```

Comment:
should be distinguish between not found paths and unprintable data?

Reply:
Possibly.  I think of the current annotation from OpenAPI is underspecified and a bit of a quick hack, so some more thought is deserved for all of this.  I was going to propose a 1.11 design extension that covered more of the edge cases - I think given server side printing depends on the CRD, not just OpenAPI, we can do a better job of making this useful.  The design goal of CRD is to invert the API model and make clients able to usefully function against a dumb api server, so a better definition (server side smarts) seems logical.